### DISCUSSION TITLE
常用 STL 使用与大模拟题

### DISCUSSION THREAD

**Discussion by SamuNatsu** [CSGrandeur/s-1problem1day1ac#482] [Algorithm]
# 常用 STL 使用方法

## 什么是 STL

STL 即 **S**tandard **T**emplate **L**ibrary —— 标准模板库，包含了大量已经打包好的实用算法与数据结构，可以在比赛过程中大幅缩减码量，提高效率。

STL 的所有内容都放在了 std 名字空间中，所以在使用 STL 函数/对象时，需要添加 `std::` 前缀或在程序开头使用 `using namespace std;` 来定位到 std 名字空间中。

当然十分不推荐直接使用 `using namespace std;`，因为这个语句直接将 std 名字空间中的所有内容释放到了当前文件环境中，会造成**名字空间污染**，你不知道什么时候就不小心误用了 std 名字空间中定义的一些函数 / 常量 / 全局变量。

所以本文将使用 `std::` 前缀作为 STL 标识。

STL 一般情况下只需要使用 `#include <bits/stdc++.h>` 引入即可（万能头文件），但是有些编译平台不支持这个头文件，所以以下内容会特意标明其原始头文件，以备需要。

同样的，部分 STL 需要 C++11 及以上标准才能使用，也会进行特殊说明。

### std::vector

- 头文件：`#include <vector>`
- 简介：变长数组容器
- 简单技术细节：通过一定手段**自动管理一段长度的连续内存空间**实现变长数组
- 用法：
	- 常用构造函数：
		1. `std::vector<类型> 变量名;`
			- 一般声明，空数组
		2. `std::vector<类型> 变量名(n);`
			- 预分配大小为 n 个元素的空间，并**自动调用该类型的无参构造函数**，同样意味着**没有默认构造函数的类型不能使用这种方法声明变长数组**
	- 常用成员方法：
		1. `operator[]`
			- 取数组下标，与一般数组的操作完全相同
			- 例：`vec[5] = 10;`
		2. `size()`
			- 返回当前数组大小
			- 例：`vec.size()`
		3. `push_back()`
			- 将一个元素插入到数组末尾
			- 例：`vec.push_back(10)`
		4. `emplace_back()` **{ C++11 }**
			- 将在数组末尾构造一个元素
			- 例：`vec.emplace_back(10)`
			- **注：** 与 `push_back()` 不同在于 `push_back()` 会调用类型的**复制构造函数**，而 `emplace_back()` 通过直接使用**有参 / 无参构造函数**，没有复制构造函数带来的性能损耗（常数优化），且在类型支持的情况下可以使用**移动构造函数**来进一步提高性能，但这就**要求该数据类型提供有参 / 无参构造函数或移动构造函数**。一般情况下推荐优先使用 `emplace_back()`。
		5. `insert()`
			- 在某个迭代器位置插入一个元素
			- 例：`vec.insert(vec.begin() + 3, 5)`
		6. `emplace()` **{ C++11 }**
			- 在某个迭代器位置构造一个元素并插入
			- 例：`vec.emplace(vec.begin() + 3, 5)`
			- **注：** 原理同 `emplace_back()`，推荐优先使用。
		7. `erase()`
			- 删除某个 / 某段迭代器位置的元素
			- 例1：`vec.erase(vec.begin() + 3)`
			- 例2：`vec.erase(vec.begin() + 1, vec.begin() + 5)`
		8. `pop_back()`
			- 删除最后一个元素
			- 例：`vec.pop_back()`
		9. `clear()`
			- 清空数组
			- 例：`vec.clear()`
		10. `front()`
			- 返回第一个元素（没啥大用，等价于 `vec[0]` ）
			- 例：`vec.front()`
		11. `back()`
			- 返回最后一个元素（这个很有用，等价于 `vec[vec.size() - 1]` ）
			- 例：`vec.back()`
		12. 常用迭代器系统
			- 正向迭代器：`begin()`、`end()`
				`begin()` 相当于数组头元素指针，**可以使用四则运算定位**。
				`end()` 是数组尾元素的**后一个**位置。
			- 反向迭代器：`rbegin()`、`rend()`
			`rbegin()` 是数组尾元素的位置。
			`rend()` 是数组头元素的**前一个**位置。
			- 使用迭代器遍历 / 访问数组本质上也是常数优化功能，`std::vector` 的迭代器直接当做指针使用即可。

### std::queue

- 头文件：`#include <queue>`
- 简介：队列容器
- 简单技术细节：只允许在头部和尾部进行处理的变长数组
- 用法：
	- 常用构造函数
		1. `std::queue<类型> 变量名;`
			- 一般声明，空队列
	- 常用成员方法
		1. `empty()`
			- 检查队列是否为空，为空返回 true
			- 例：`que.empty()`
			- 注：`std::queue` 没有 `clear()` 方法，所以一般通过不停检查队列非空并弹出栈顶来清空栈
		2. `size()`
			- 返回队列当前含有的元素数量
			- 例：`que.size()`
		3. `push()`
			- 向队尾插入一个元素
			- 例：`que.push(810)`
		4. `emplace()` **{ C++11 }**
			- 在队尾构造一个元素
			- 例：`que.emplace(810)`
			- **注：** 所有 emplace 类函数的原理都是相同的，以下不再赘述。
		5. `front()`
			- 获得队首元素，但是**不弹出该元素**
			- 例：`que.front()`
		6. `back()`
			- 获得队尾元素，但是**不弹出该元素**（没啥大用）
			- 例：`que.back()`
		7. `pop()`
			- 从队首弹出一个元素，但是**不返回该元素**
			- 例：`que.pop()`
			

### std::deque

- 头文件：`#include <deque>`
- 简介：双端队列
- 简单技术细节：可以在头部和尾部进行更多操作的变长数组，与 `std::vector` 可以互相替换
- 用法：
	- 常用构造函数
		与 `std::vector` 相同。
	- 常用成员方法
		1. `std::vector` 的所有常用成员方法 `std::deque` 都能使用。
		2. `push_front()`
			- 将一个元素插入到队首
			- 例：`deq.push_front(10)`
		3. `emplace_front()` **{ C++11 }**
			- 在队首构造一个元素
			- 例：`deq.emplace_front(10)`
		4. `pop_front()`
			- 弹出队首元素，但是**不返回该元素**
			- 例：`deq.pop_front()`

### std::stack

- 头文件：`#include <stack>`
- 简介：栈容器
- 简单技术细节：只能在尾部进行处理的变长数组
- 用法：
	- 常用构造函数
		1. `std::stack<类型> 变量名;`
			- 一般声明，空栈
	- 常用成员方法
		1. `push()`
			- 将一个元素压入栈
			- 例：`stk.push(10)`
		2. `emplace()` **{ C++11 }**
			- 在栈顶构造一个元素
			- 例：`stk.emplace(10)`
		3. `top()`
			- 获得栈顶元素，但是不弹出
			- 例：`stk.top()`
		4. `pop()`
			- 弹出栈顶元素，但是不返回
			- 例：`stk.pop()`
		5. `empty()`
			- 检查栈是否为空
			- 例：`stk.empty()`
			- 注：`std::stack` 没有 `clear()` 方法，所以一般通过不停检查栈非空并弹出栈顶来清空栈

### std::priority_queue

- 头文件：`#include <queue>`
- 简介：优先队列（大根堆）
- 简单技术细节：在变长数组上维护了一个二叉堆以实现优先级访问
- 用法：
	- 常用构造函数
		1. `std::priority_queue<类型> 变量名;`
			- 一般声明，空优先队列
			- **注：类型必须是可比较的**（实现了小于运算符）
	- 常用成员方法
		1. `push()`
			- 将一个元素放入队列
			- 例：`pque.push(10)`
		2. `emplace()` **{ C++11 }**
			- 在队列中构造一个元素
			- 例：`pque.emplace(10)`
		3. `top()`
			- 获得最高优先级元素，但是不弹出
			- 例：`pque.top()`
		4. `pop()`
			- 弹出最高优先级元素，但是不返回
			- 例：`pque.pop()`
		5. `empty()`
			- 检查队列是否为空
			- 例：`pque.empty()`
			- 注：`std::priority_queue` 没有 `clear()` 方法，所以一般通过不停检查非队列空并弹出来清空队列

### std::map

- 头文件：`#include <map>`
- 简介：映射表（字典，键值对表）
- 简单技术细节：使用红黑树（二叉平衡搜索树）实现平均 `O(logn)` 复杂度操作
- 用法：
	- 常用构造函数
		1. `std::map<键类型, 值类型> 变量名;`
			- 一般声明，空映射表
			- **注：键类型需要是可比较的数据结构**（实现了小于和**等于**运算符）
	- 常用成员方法
		1. `operator[]`
			- 取下标，获得键对应的值
			- 例：`mp["abc"] = 123`
		2. `clear()`
			- 清空表
			- 例：`mp.clear()`
		3. `count()`
			- 检查是否存在键值，存在返回 1，否则为 0
			- 例：`mp.count("abc")`
		4. `erase()`
			- 根据键值删除一个元素
			- 例：`mp.erase("abc")`
		5. `lower_bound()`
			- 返回第一个不小于指定键值的元素迭代器（二分查找）
			- 例：`mp.lower_bound("c")`
		6. `upper_bound()`
			- 返回第一个大于指定键值的元素迭代器（二分查找）
			- 例：`mp.upper_bound("a")`
		7. 常用迭代器系统
			- 正向迭代器：`begin()`、`end()`
				`begin()` 相当于最小元素的指针，**不可以使用四则运算定位**。
				`end()` 是最大元素的**后一个**位置（第一个非法位置）。
			- 反向迭代器：`rbegin()`、`rend()`
			- `std::map` 的迭代器只能使用自增（++）和自减（--）运算移动

### std::unordered_map

- 头文件：`#include <unordered_map>`
- 简介：哈希表（无序字典）
- 简单技术细节：使用可变长数组实现哈希表结构，完成平均 `O(1)` 的操作
- 用法：
	- 常用构造函数
		1. `std::unordered_map<键类型, 值类型> 变量名;`
			- 一般声明，空哈希表。
			- **注：键类型需要是可哈希数据结构**（需要实现类型的等于运算符以及类型对应的哈希类）
	- 常用成员方法
		1. 大部分与 `std::map` 相同，但是没有 `lower_bound()` 和 `upper_bound()`

### std::set

- 头文件：`#include <set>`
- 简介：有序集合
- 简单技术细节：与 `std::map` 一样
- 用法：
	- 常用构造函数
		1. `std::set<类型> 变量名;`
			- 一般声明，空集合
			- **注：类型需要是可比较的数据结构**（实现小于和**等于**运算符）
	- 常用成员方法
		1. `insert()`
			- 插入一个元素
			- 例：`st.insert(10)`
		2. `emplace()` **{ C++11 }**
			- 在集合中构造一个元素
			- 例：`st.emplace(10)`
		3. `count()`
			- 检查是否存在值，存在返回 1，否则为 0
			- 例：`st.count(10)`
		4. `erase()`
			- 根据值删除一个元素
			- 例：`st.erase(10)`
		5. `lower_bound()`
			- 返回第一个不小于指定值的元素迭代器（二分查找）
			- 例：`st.lower_bound(10)`
		6. `upper_bound()`
			- 返回第一个大于指定值的元素迭代器（二分查找）
			- 例：`st.upper_bound(10)`
		7. 常用迭代器系统
			- 与 `std::map` 一样

### std::unordered_set

- 头文件：`#include <unordered_set>`
- 简介：无序集合
- 简单技术细节：与 `std::unordered_map` 一样
- 用法：
	- 常用构造函数
		1. `std::unordered_set<类型> 变量名;`
			- 一般声明，空集合
			- **注：类型需要是可哈希的数据结构**（需要实现类型的等于运算符以及类型对应的哈希类）
	- 常用成员方法
		1. 大部分与 `std::unordered_set` 相同，但是没有 `lower_bound()` 和 `upper_bound()`

### std::stringstream
- 头文件：`#include <sstream>`
- 简介：字符串流，用来构造字符串或转化字符串为数据
- 简单技术细节：内部使用变长字符串充当缓冲区完成输入输出操作
- 用法：
	- 常用构造函数
		1. `std::stringstream 变量名;`
			- 一般声明，空字符串流
	- 常用成员方法
		1. `operator>>`
			- 将流中的数据输出
			- 例：`ss >> a`
		2. `operator<<`
			- 向流输入数据
			- 例：`ss << 1234`
		3. `str()`
			- 获得流中的内容或设置流内容
			- 例1：`ss.str()`
			- 例2：`ss.str("12345")`

## 常用算法函数

STL 算法使用的头文件都是 `#include <algorithm>`。

### std::sort
- 快速排序函数
- 例：`std::sort(a, a + N)`

### std::lower_bound
- 二分查找第一个不小于的元素（要求查找序列有序）
- 例：`std::lower_bound(a, a + N, 123)`

### std::upper_bound
- 二分查找第一个大于的元素（要求查找序列有序）
- 例：`std::upper_bound(a, a + N, 123)`

### std::to_string
- 将数字转换成字符串
- 例：`std::to_string(1234)`

### std::stoi, std::stol, std::stoll
- 将字符串转换成数字（int/long/long long）
- 例：`std::stoi("114514")`

### std::unique
- 将有序数列中重复的元素放到数列末尾，返回去重后的数列长度
- 例：`std::unique(a, a + N)`

## 模拟题

所有没有任何算法要素存在的题目都可以统称为**模拟题**，因此模拟题考查的是学生对于现实问题的解决能力，简单来说就是**信息界的看图写话**。

模拟题没有固定的规律 / 算法 / 思路可言，只能依靠长期的做题和做项目来提高。且一般模拟代码的码量很高，题目又臭又长，读题时间和实现时间有时能五五开，所以需要在平时**锻炼快速审题能力**，并合理运用 STL 这个 C++ 原生提供的工具**减少代码实现时间**。

**Comment by TUStarry** [comment]
今日习题【记分】
1. 201312-2   ISBN号码
2. 202009-2   风险人群筛查
3. 202006-2   稀疏向量
4. 201803-2   碰撞的小球
5. 201712-2   游戏

可在[CSP官网](https://www.cspro.org/)点击“模拟考试”或在[ACWing](https://www.acwing.com/problem)搜索对应题目练习。

**Comment by TUStarry** [comment]
今日习题【不计分】
1. 201912-2   回收站选址
2. 201909-2   小明种苹果（续）
3. 201903-2   二十四点
4. 201812-2   小明放学
5. 201809-2   买菜
6. 201709-3   JSON查询
7. 201909-3   字符画
8. 201703-3   Markdown

可在[CSP官网](https://www.cspro.org/)点击“模拟考试”或在[ACWing](https://www.acwing.com/problem)搜索对应题目练习。
